{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"LAST_homeolog.ipynb","provenance":[],"collapsed_sections":[],"authorship_tag":"ABX9TyPrgcmlPF5DMWeEtbrimzVB"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"markdown","source":["# Código para obtener las tétradas de homeólogos del genoma de _Fragaria x ananassa_"],"metadata":{"id":"kwaEaJlACQlA"}},{"cell_type":"code","execution_count":null,"metadata":{"id":"lLPHqgcboCa1"},"outputs":[],"source":["#Download LAST from repository and installation\n","!wget https://gitlab.com/mcfrith/last/-/archive/1256/last-1256.zip\n","!unzip last-1256.zip\n","!cd last-1256; make install"]},{"cell_type":"code","source":["#Download the CDS of each subgenome and decompress them\n","!wget https://github.com/agm-bioinfo/Fragaria-x-ananassa-RNA-seq/raw/main/Homeologs/A.zip\n","!wget https://github.com/agm-bioinfo/Fragaria-x-ananassa-RNA-seq/raw/main/Homeologs/B.zip\n","!wget https://github.com/agm-bioinfo/Fragaria-x-ananassa-RNA-seq/raw/main/Homeologs/C.zip\n","!wget https://github.com/agm-bioinfo/Fragaria-x-ananassa-RNA-seq/raw/main/Homeologs/D.zip\n","\n","!unzip A.zip\n","!unzip B.zip\n","!unzip C.zip\n","!unzip D.zip"],"metadata":{"id":"HxeCETe3Cx9s"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#Create databases from each subgenome\n","!lastdb -uNEAR -R01 -P0 A_db A.fa\n","!lastdb -uNEAR -R01 -P0 B_db B.fa\n","!lastdb -uNEAR -R01 -P0 C_db C.fa\n","!lastdb -uNEAR -R01 -P0 D_db D.fa"],"metadata":{"id":"7pL5qGoxc9Bl"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#Align each subgenome to every other subgenome\n","#Not all alignments are necesary but can be useful if you have gff files and sequences coordinates and need them\n","#cause alignments A_vs_B and B_vs_A will report diferent coordinates based on the first database\n","!lastal A_db -P0 -D10000000000 B.fa | last-map-probs -m 0.49 > A.B.maf\n","!lastal A_db -P0 -D10000000000 C.fa | last-map-probs -m 0.49 > A.C.maf\n","!lastal A_db -P0 -D10000000000 D.fa | last-map-probs -m 0.49 > A.D.maf\n","\n","!lastal B_db -P0 -D10000000000 A.fa | last-map-probs -m 0.49 > B.A.maf\n","!lastal B_db -P0 -D10000000000 C.fa | last-map-probs -m 0.49 > B.C.maf\n","!lastal B_db -P0 -D10000000000 D.fa | last-map-probs -m 0.49 > B.D.maf\n","\n","!lastal C_db -P0 -D10000000000 B.fa | last-map-probs -m 0.49 > C.B.maf\n","!lastal C_db -P0 -D10000000000 A.fa | last-map-probs -m 0.49 > C.A.maf\n","!lastal C_db -P0 -D10000000000 D.fa | last-map-probs -m 0.49 > C.D.maf\n","\n","!lastal D_db -P0 -D10000000000 B.fa | last-map-probs -m 0.49 > D.B.maf\n","!lastal D_db -P0 -D10000000000 C.fa | last-map-probs -m 0.49 > D.C.maf\n","!lastal D_db -P0 -D10000000000 D.fa | last-map-probs -m 0.49 > D.A.maf"],"metadata":{"id":"PP4Owr2TtQ9U"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#Download scripts from https://github.com/tony-kuo/eagle/tree/master/scripts to get best reciprocal hits between alignments and generate groups of 4 homeologs from each subgenome\n","!wget https://raw.githubusercontent.com/tony-kuo/eagle/master/scripts/homeolog_genotypes.py\n","!wget https://raw.githubusercontent.com/tony-kuo/eagle/master/scripts/quad_homeolog.pl"],"metadata":{"id":"GYl6jR5NbloT"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#Generate best hits between the alignments\n","!python homeolog_genotypes.py -o A.vs.B A.B.maf B.A.maf\n","!python homeolog_genotypes.py -o A.vs.C A.C.maf C.A.maf\n","!python homeolog_genotypes.py -o A.vs.D A.D.maf D.A.maf\n","\n","!python homeolog_genotypes.py -o B.vs.A B.A.maf A.B.maf\n","!python homeolog_genotypes.py -o B.vs.C B.C.maf C.B.maf\n","!python homeolog_genotypes.py -o B.vs.D B.D.maf D.B.maf\n","\n","!python homeolog_genotypes.py -o C.vs.B C.B.maf B.C.maf\n","!python homeolog_genotypes.py -o C.vs.A C.A.maf A.C.maf\n","!python homeolog_genotypes.py -o C.vs.D C.D.maf D.C.maf\n","\n","!python homeolog_genotypes.py -o D.vs.B D.B.maf B.D.maf\n","!python homeolog_genotypes.py -o D.vs.C D.C.maf C.D.maf\n","!python homeolog_genotypes.py -o D.vs.A D.A.maf A.D.maf"],"metadata":{"id":"TNfdx2Fwnx8M"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["!perl quad_homeolog.pl A.vs.B.reciprocal_best A.vs.C.reciprocal_best A.vs.D.reciprocal_best B.vs.C.reciprocal_best B.vs.D.reciprocal_best C.vs.D.reciprocal_best > homeolog.ABCD.list"],"metadata":{"id":"kSCnUa9IhHoq"},"execution_count":null,"outputs":[]}]}